compiler-core/src/package_interface.rs

```diff
@@ -0,0 +1,280 @@
+                        deprecation: Deprecation::NotDeprecated,
+                        documentation: documentation.clone(),
+                    },
+                );
+            }
+
+            // A public top-level function.
+            Definition::Function(Function {
+                public: true,
+                name,
+                arguments,
+                deprecation,
+                return_type,
+                documentation,
+                supported_targets,
+                location: _,
+                end_position: _,
+                body: _,
+                return_annotation: _,
+                external_erlang: _,
+                external_javascript: _,
+            }) => {
+                let _ = values.insert(
+                    name.clone(),
+                    ValueInterface {
+                        supported_targets: supported_targets.to_vec(),
+                        deprecation: deprecation.clone(),
+                        documentation: documentation.clone(),
+                        type_: TypeInterface::Fn {
+                            parameters: arguments
+                                .iter()
+                                .map(|arg| TypeInterface::from_type(arg.type_.as_ref()))
+                                .collect(),
+                            return_: Box::new(TypeInterface::from_type(return_type)),
+                        },
+                    },
+                );
+            }
+
+            // Private definitions are not included.
+            Definition::Function(_) => {}
+            Definition::CustomType(_) => {}
+            Definition::ModuleConstant(_) => {}
+
+            // Type aliases are ignored since those are also erased from the
+            // generated documentation.
+            Definition::TypeAlias(_) => {}
+
+            // Imports are ignored.
+            Definition::Import(_) => {}
+        }
+    }
+    (types, values)
+}
+
+impl TypeInterface {
+    fn from_type(type_: &Type) -> TypeInterface {
+        match type_ {
+            Type::Fn { args, retrn } => TypeInterface::Fn {
+                parameters: args
+                    .iter()
+                    .map(|arg| TypeInterface::from_type(arg.as_ref()))
+                    .collect(),
+                return_: Box::new(TypeInterface::from_type(retrn)),
+            },
+
+            Type::Tuple { elems } => TypeInterface::Tuple {
+                elements: elems
+                    .iter()
+                    .map(|elem| TypeInterface::from_type(elem.as_ref()))
+                    .collect(),
+            },
+
+            Type::Var { type_ } => match type_.as_ref().try_borrow() {
+                // Package serialisation happens after inference has taken place
+                // At this point no one should be helding a mutable reference of
+                // the type: the borrow should never fail.
+                Err(_) => unreachable!(),
```

Comment:
Use `expect` here, that's how we make assertions about things that must hold true and otherwise be a fatal error.